test/neural_de_lux.jl

```diff
@@ -234,7 +234,7 @@ grads = Zygote.gradient(()->sum(sode(xs)),Flux.params(xs,sode))
 @test ! iszero(grads[sode.p])
 @test ! iszero(grads[sode.p][end])
 
-sode = NeuralDSDE(luxdudt,luxdudt2,(0.0f0,.1f0),LambaEulerHeun(),saveat=0.0:0.01:0.1)
+sode = NeuralDSDE(luxdudt,luxdudt2,(0.0f0,.1f0),EulerHeun(),saveat=0.0:0.01:0.1,dt=0.1)
```

Comment:
why does a fixed `dt` make sense?

Reply:
The adaptive scheme was throwing the dtmin warning that says aborting. This one was working flawlessly for vector input. This too is failing for the matrix with some mul! error I'm looking into it rn

Reply:
Oh, I should've made that a dt of 0.01

Reply:
+1 on the `mul!` error. There seems to be some dimension mismatch problem. I worked around it in my research codebases using:

```julia
# This is a weird type piracy but is needed to make the NeuralDSDE to work
function LinearAlgebra.mul!(A::AbstractVector{T}, B::AbstractMatrix{T},
                            C::AbstractMatrix{T}, b1::Bool, b2::Bool) where {T}
  LinearAlgebra.mul!(reshape(A, :, 1), B, reshape(C, :, 1), b1, b2)
  return A
end
```

Reply:
But why is it there?

Reply:
Also [this](https://github.com/SciML/StochasticDiffEq.jl/blob/3d56e689a14d759e5ad0ac7eddc2cd0bc4b0bab3/src/perform_step/low_order.jl#L92) `mul!` is hitting the dispatch defined at [LeastSquaresOptim](https://github.com/matthieugomez/LeastSquaresOptim.jl/blob/main/src/solver/iterative_lsmr.jl) dispatch for some reason

Reply:
```julia
Error During Test at /home/runner/.julia/packages/SafeTestsets/A83XK/src/SafeTestsets.jl:25
  Got exception outside of a @test
  LoadError: MethodError: no method matching mul!(::Vector{Float32}, ::Matrix{Float32}, ::Matrix{Float64}, ::Bool, ::Bool)
  Closest candidates are:
    mul!(::Any, ::LeastSquaresOptim.PreconditionedMatrix{TA, Tp, Tx}, ::Any, ::Number, ::Number) where {TA, Tp, Tx} at ~/.julia/packages/LeastSquaresOptim/B1WKJ/src/solver/iterative_lsmr.jl:30
    mul!(::Any, ::LeastSquaresOptim.DampenedMatrix, ::Any, ::Number, ::Number) where T at ~/.julia/packages/LeastSquaresOptim/B1WKJ/src/solver/iterative_lsmr.jl:87
    mul!(::Any, ::LeastSquaresOptim.MyAdjoint{LeastSquaresOptim.PreconditionedMatrix{TA, Tp, Tx}}, ::Any, ::Number, ::Number) where {TA, Tp, Tx} at ~/.julia/packages/LeastSquaresOptim/B1WKJ/src/solver/iterative_lsmr.jl:36
```

Reply:
That's not hitting the LeastSquaresOptim dispatches, it's just showing that those exist. The issue is that the last term should be `Matrix{Float32}`. Where is the non-diagonal noise matrix defined? I would assume it would be this code.

Reply:
Is it coming from here https://github.com/SciML/SciMLSensitivity.jl/blob/master/src/concrete_solve.jl#L733 ? Are the parameters Float64?

Reply:
It being `Matrix{Float64}` is a part of the problem, but also two matrices are being passed to mul! to give a vector. I just converted `W.dW` directly to run:

```julia
ERROR: MethodError: no method matching mul!(::Vector{Float32}, ::Matrix{Float32}, ::Matrix{Float32}, ::Bool, ::Bool)
Closest candidates are:
  mul!(::Any, !Matched::LeastSquaresOptim.MyAdjoint{LeastSquaresOptim.PreconditionedMatrix{TA, Tp, Tx}}, ::Any, ::Number, ::Number) where {TA, Tp, Tx} at C:\Users\abhishek_bhatt\.julia\packages\LeastSquaresOptim\B1WKJ\src\solver\iterative_lsmr.jl:36
  mul!(::Any, !Matched::LeastSquaresOptim.MyAdjoint{LeastSquaresOptim.DampenedMatrix{TA, Tx}}, ::Any, ::Number, ::Number) where {TA, Tx} at C:\Users\abhishek_bhatt\.julia\packages\LeastSquaresOptim\B1WKJ\src\solver\iterative_lsmr.jl:95
  mul!(::Any, !Matched::LeastSquaresOptim.PreconditionedMatrix{TA, Tp, Tx}, ::Any, ::Number, ::Number) where {TA, Tp, Tx} at C:\Users\abhishek_bhatt\.julia\packages\LeastSquaresOptim\B1WKJ\src\solver\iterative_lsmr.jl:30
```

Reply:
Change it to `LinearAlgebra.mul!`? Is that package shadowing `mul!`?

Reply:
Why is that package in the image at all BTW? What even uses LeastSquaresOptim?

Reply:
That's what I was wondering, none of the packages involved depend on it

Reply:
> The issue is that the last term should be Matrix{Float32}.

I don't think that is the problem. It is the `Matrix x Matrix` = `Vector` that is causing the error.

```julia
using LinearAlgebra

A = zeros(2); B = ones(2, 2); C = ones(2, 1);

LinearAlgebra.mul!(A, B, C)
# ERROR: MethodError: no method matching mul!(::Vector{Float64}, ::Matrix{Float64}, ::Matrix{Float64}, ::Bool, ::Bool)
# Closest candidates are:
#   mul!(::StridedVector{T}, ::StridedVecOrMat{T}, ::StridedVector{T}, ::Number, ::Number) where T<:Union{Float32, Float64, ComplexF32, ComplexF64} at /mnt/julia/juliaup/julia-1.8.2+0.x64/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:65
#   mul!(::StridedMatrix{T}, ::StridedVecOrMat{T}, ::StridedVecOrMat{T}, ::Number, ::Number) where T<:Union{Float32, Float64, ComplexF32, ComplexF64} at /mnt/julia/juliaup/julia-1.8.2+0.x64/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:159
#   mul!(::StridedVecOrMat, ::SymTridiagonal, ::StridedVecOrMat, ::Number, ::Number) at /mnt/julia/juliaup/julia-1.8.2+0.x64/share/julia/stdlib/v1.8/LinearAlgebra/src/tridiag.jl:217
#   ...
# Stacktrace:
#  [1] mul!(C::Vector{Float64}, A::Matrix{Float64}, B::Matrix{Float64})
#   @ LinearAlgebra /mnt/julia/juliaup/julia-1.8.2+0.x64/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:276
#  [2] top-level scope
#   @ REPL[3]:1
```